llvm/lib/Transforms/InstCombine/InstCombineMulDivRem.cpp

```diff
@@ -626,6 +626,150 @@ Instruction *InstCombinerImpl::foldPowiReassoc(BinaryOperator &I) {
+    // changes to R1 and R2 are anyway done as part of other transforms.R1 and
+    // R2 must either be global or must have single local use.
+    if (!X->hasNUsesOrMore(4))
+      return false;
+    // check if reciprocalFP is enabled
+    bool RecipFPMath = dyn_cast<FPMathOperator>(X)->hasAllowReciprocal();
+    if (!RecipFPMath)
+      return false;
+    if (IsStrictFP(X))
+      return false;
+    return true;
+  };
+  if (!XConstraintsSatisfied())
+    return false;
+
+  // check the constraints on instructions in R1
+  auto R1ConstraintsSatisfied = [BBr1, &IsStrictFP](Instruction *I) {
+    // when you have multiple instructions residing in R1 and R2 respectively,
+    // its difficult to generate combination of (R1,R2) and then check if we
+    // have the required pattern. So, for now, just be conservative.
+    if (I->getParent() != BBr1)
+      return false;
+    if (!I->hasNUsesOrMore(1))
+      return false;
+    if (IsStrictFP(I))
+      return false;
+    // The optimization tries to convert
+    // R1 = div * div    where, div = 1/sqrt(a)
+    // to
+    // R1 = 1/a
+    // Now, this simplication does not work because sqrt(a)=NaN when a<0
+    if (!I->hasNoNaNs())
+      return false;
+    // sqrt(-0.0) = -0.0, and doing this simplication would change the sign of
+    // the result.
+    if (!I->hasNoSignedZeros())
+      return false;
+    return true;
+  };
+  if (!std::all_of(R1.begin(), R1.end(), R1ConstraintsSatisfied))
+    return false;
+
+  // check the constraints on instructions in R2
+  auto R2ConstraintsSatisfied = [BBr2, &IsStrictFP](Instruction *I) {
+    // when you have multiple instructions residing in R1 and R2 respectively,
+    // its
+    // difficult to generate combination of (R1,R2) and then check if we have
+    // the required pattern. So, for now, just be conservative.
+    if (I->getParent() != BBr2)
+      return false;
+    if (!I->hasNUsesOrMore(1))
+      return false;
+    if (IsStrictFP(I))
+      return false;
+    // This simplication changes
+    // R2 = a * 1/sqrt(a)
+    // to
+    // R2 = sqrt(a)
+    // Now, sqrt(-0.0) = -0.0 and doing this simplication would produce -0.0
+    // instead of NaN.
+    if (!I->hasNoSignedZeros())
+      return false;
+    return true;
+  };
+  if (!std::all_of(R2.begin(), R2.end(), R2ConstraintsSatisfied))
+    return false;
+
+  auto XR1R2ConstraintsSatisfied = [=]() {
+    const Function *F = X->getFunction();
+    bool UnsafeFPMath = F->getFnAttribute("unsafe-fp-math").getValueAsBool();
+    if (!UnsafeFPMath)
+      return false;
+    if (!BBConstraintsSatisfied())
+      return false;
+    return true;
+  };
+  if (!XR1R2ConstraintsSatisfied())
+    return false;
+
+  return true;
+}
+
+void getFSqrtDivOptPattern(Value *Div, SmallSetVector<Instruction *, 2> &R1,
+                           SmallSetVector<Instruction *, 2> &R2) {
+  Value *A;
+  if (match(Div, m_FDiv(m_FPOne(), m_Sqrt(m_Value(A))))) {
+    for (auto U : Div->users()) {
```

Comment:
Should this try only looking at the 2 expected users, instead of iterating arbitrary users?

Reply:
yes, will refine this